react/features/calendar-sync/web/microsoftCalendar.js

```diff
@@ -0,0 +1,531 @@
+/* @flow */
+
+import rs from 'jsrsasign';
+import { Client } from '@microsoft/microsoft-graph-client';
+
+import { CALENDAR_API_STATES } from '../constants';
+import {
+    setCalendarAPIAuthState,
+    setCalendarAPIState,
+    setCalendarProfileEmail
+} from '../actions';
+
+/**
+ * The Microsoft API scopes to request access for calendar.
+ *
+ * @type {Array<string>}
+ */
+const MS_API_SCOPES = 'openid profile Calendars.Read';
+
+/**
+ * The redirect URL to be used by the Microsoft API.
+ * @type {string}
+ */
+const REDIRECT_URI = `${window.location.origin}/static/msredirect.html`;
+
+/**
+ * The URL to use when authenticating using Microsoft API.
+ * @type {string}
+ */
+const AUTH_ENDPOINT
+    = 'https://login.microsoftonline.com/common/oauth2/v2.0/authorize?';
+
+const logger = require('jitsi-meet-logger').getLogger(__filename);
+
+/**
+ * Loads and interacts with the Microsoft Calendar API.
+ */
+export class MicrosoftCalendarApi {
+    /**
+     * The ID for the Microsoft client application used to init the API.
+     */
+    _appClientID: string;
+
+    /**
+     * The current request data.
+     */
+    _currentRequest: ?Object;
+
+    /**
+     * The last valid response data.
+     */
+    _lastValidResponse: ?Object;
+
+    /**
+     * The popup window used for authentication, using the instance to close
+     * the popup once we have all the info we need.
+     */
+    _popupAuthWindow: ?Object;
+
+    /**
+     * The redux {@code dispatch} function.
+     */
+    _dispatch: Dispatch<*>;
+
+    /**
+     * The redux function that gets/retrieves the redux state.
+     */
+    _getState: Function;
+
+    /**
+     * Initializes a new Microsoft Calendar API instance.
+     *
+     * @param {string} appClientID - The ID for the Microsoft client application
+     * used to init the API.
+     * @param {Object} store - The redux store.
+     */
+    constructor(appClientID: string, store: Object) {
+        this._appClientID = appClientID;
+        this._dispatch = store.dispatch;
+        this._getState = store.getState;
+        this._currentRequest = {};
+
+        // checks post messages for events from the popup
+        window.addEventListener('message', this._onPostMessage.bind(this));
```

Comment:
If switching between microsoft and google, would this listener cause any problems? Do you need to do any cleanup in a deconstructor/destroy method?

Reply:
I don't see how it can cause any problems, but cleaning is something we should take care. I will check it now.